File: packages/core/src/protocol.ts
import { Schema } from "effect"
import { PromptChip } from "./prompt.js"
import { AgentQuestion, AskUserQuestionAnswer } from "./questions.js"

export const StopReasonView = Schema.Literals([
  "end_turn",
  "cancelled",
  "max_tokens",
  "max_turn_requests",
  "refusal",
  "unknown",
])

export type StopReasonView = typeof StopReasonView.Type

export const displayStopReason = (reason: string): StopReasonView => {
  const decoded = Schema.decodeUnknownExit(StopReasonView)(reason)
  return decoded._tag === "Success" ? decoded.value : "unknown"
}

const PermissionOption = Schema.Struct({
  optionId: Schema.String,
  name: Schema.String,
  kind: Schema.String,
})

const ToolRow = Schema.Struct({
  id: Schema.String,
  title: Schema.String,
  kind: Schema.String,
  status: Schema.String,
  additions: Schema.optionalKey(Schema.Number),
  deletions: Schema.optionalKey(Schema.Number),
  input: Schema.optionalKey(Schema.String),
  output: Schema.optionalKey(Schema.String),
})

const Occupancy = Schema.Struct({
  used: Schema.Number,
  size: Schema.Number,
})

export const ReasoningChoice = Schema.Struct({
  value: Schema.String,
  name: Schema.String,
})

export type ReasoningChoice = typeof ReasoningChoice.Type

const SessionReasoning = Schema.Struct({
  id: Schema.String,
  current: Schema.String,
  options: Schema.Array(ReasoningChoice),
})

const ModelReasoning = Schema.Struct({
  current: Schema.String,
  options: Schema.Array(ReasoningChoice),
})

const MentionFile = Schema.Struct({
  path: Schema.String,
  absPath: Schema.String,
})

const SlashCommand = Schema.Struct({
  name: Schema.String,
  description: Schema.String,
  hint: Schema.optionalKey(Schema.String),
})

export const SessionModeOption = Schema.Struct({
  id: Schema.String,
  name: Schema.String,
})

export type SessionModeOption = typeof SessionModeOption.Type

export const SessionModelOption = Schema.Struct({
  modelId: Schema.String,
  name: Schema.String,
  description: Schema.optionalKey(Schema.String),
  contextWindow: Schema.optionalKey(Schema.Number),
  reasoning: Schema.optionalKey(ModelReasoning),
})

export type SessionModelOption = typeof SessionModelOption.Type

export const HostMsg = Schema.Union([
  Schema.TaggedStruct("ready", {}),
  Schema.TaggedStruct("sessionMeta", {
    sessionId: Schema.String,
    title: Schema.String,
    modeId: Schema.String,
    modelId: Schema.optionalKey(Schema.String),
    occupancy: Schema.optionalKey(Occupancy),
    reasoning: Schema.optionalKey(SessionReasoning),
    availableModes: Schema.optionalKey(Schema.Array(SessionModeOption)),
    availableModels: Schema.optionalKey(Schema.Array(SessionModelOption)),
  }),
  Schema.TaggedStruct("userMessage", {
    turnId: Schema.String,
    text: Schema.String,
    chips: Schema.Array(PromptChip),
    steer: Schema.optionalKey(Schema.Boolean),
  }),
  Schema.TaggedStruct("agentChunk", {
    turnId: Schema.String,
    messageId: Schema.optionalKey(Schema.String),
    text: Schema.String,
  }),
  Schema.TaggedStruct("thoughtChunk", {
    turnId: Schema.String,
    text: Schema.String,
  }),
  Schema.TaggedStruct("toolGroup", {
    turnId: Schema.String,
    tools: Schema.Array(ToolRow),
  }),
  Schema.TaggedStruct("permissionCard", {
    requestId: Schema.String,
    toolCallId: Schema.String,
    title: Schema.String,
    options: Schema.Array(PermissionOption),
    hasDiff: Schema.Boolean,
  }),
  Schema.TaggedStruct("planCard", {
    requestId: Schema.String,
    planMarkdown: Schema.String,
  }),
  Schema.TaggedStruct("questionCard", {
    requestId: Schema.String,
    questions: Schema.Array(AgentQuestion),
  }),
  Schema.TaggedStruct("elicitCard", {
    requestId: Schema.String,
    serverName: Schema.String,
    mode: Schema.Literals(["form", "url"]),
    title: Schema.String,
    url: Schema.optionalKey(Schema.String),
  }),
  Schema.TaggedStruct("availableCommands", {
    commands: Schema.Array(SlashCommand),
  }),
  Schema.TaggedStruct("mentionResults", {
    query: Schema.String,
    files: Schema.Array(MentionFile),
  }),
  Schema.TaggedStruct("composerChip", {
    path: Schema.String,
    absPath: Schema.String,
    startLine: Schema.optionalKey(Schema.Number),
    endLine: Schema.optionalKey(Schema.Number),
    languageId: Schema.optionalKey(Schema.String),
    excerpt: Schema.optionalKey(Schema.String),
    source: Schema.Literals(["selection", "file", "active", "mention"]),
  }),
  Schema.TaggedStruct("turnEnd", {
    turnId: Schema.String,
    stopReason: StopReasonView,
  }),
  Schema.TaggedStruct("queued", {
    count: Schema.Number,
  }),
  Schema.TaggedStruct("changesSummary", {
    fileCount: Schema.Number,
    additions: Schema.Number,
    deletions: Schema.Number,
  }),
  Schema.TaggedStruct("editorContext", {
    path: Schema.optionalKey(Schema.String),
    startLine: Schema.optionalKey(Schema.Number),
    startCol: Schema.optionalKey(Schema.Number),
    endLine: Schema.optionalKey(Schema.Number),
    endCol: Schema.optionalKey(Schema.Number),
    hasSelection: Schema.Boolean,
    excerpt: Schema.optionalKey(Schema.String),
  }),
  Schema.TaggedStruct("settingsState", {
    cliPath: Schema.String,
    nodePath: Schema.String,
    includeActiveFileByDefault: Schema.Boolean,
    useCtrlEnterToSend: Schema.Boolean,
    changesPresentation: Schema.Literals(["toast", "pane"]),
  }),
  Schema.TaggedStruct("mcpCatalog", {
    loading: Schema.Boolean,
    healthyCount: Schema.Number,
    failingCount: Schema.Number,
    servers: Schema.Array(Schema.Struct({
      name: Schema.String,
      transport: Schema.String,
      source: Schema.String,
      healthy: Schema.Boolean,
      toolCount: Schema.optionalKey(Schema.Number),
      tools: Schema.optionalKey(Schema.Array(Schema.Struct({
        name: Schema.String,
        enabled: Schema.Boolean,
        description: Schema.optionalKey(Schema.String),
      }))),
      checks: Schema.Array(Schema.Struct({
        label: Schema.String,
        passed: Schema.Boolean,
        detail: Schema.optionalKey(Schema.String),
        hint: Schema.optionalKey(Schema.String),
      })),
    })),
    error: Schema.optionalKey(Schema.String),
  }),
  Schema.TaggedStruct("error", {
    code: Schema.optionalKey(Schema.String),
    message: Schema.String,
  }),
  Schema.TaggedStruct("clearTranscript", {}),
  Schema.TaggedStruct("restoreTranscript", {
    messages: Schema.Array(Schema.Unknown),
  }),
])

export type HostMsg = typeof HostMsg.Type

export const WebviewMsg = Schema.Union([
  Schema.TaggedStruct("ready", {}),
  Schema.TaggedStruct("send", {
    text: Schema.String,
    chips: Schema.Array(PromptChip),
  }),
  Schema.TaggedStruct("queue", {
    text: Schema.String,
    chips: Schema.Array(PromptChip),
  }),
  Schema.TaggedStruct("steer", {
    text: Schema.String,
    chips: Schema.Array(PromptChip),
  }),
  Schema.TaggedStruct("cancel", {}),
  Schema.TaggedStruct("permissionChoice", {
    requestId: Schema.String,
    optionId: Schema.String,
  }),
  Schema.TaggedStruct("permissionPark", {
    requestId: Schema.String,
  }),
  Schema.TaggedStruct("openDiff", {
    requestId: Schema.String,
  }),
  Schema.TaggedStruct("planVerdict", {
    requestId: Schema.String,
    verdict: Schema.Literals(["approved", "cancelled", "abandoned"]),
    comment: Schema.optionalKey(Schema.String),
  }),
  Schema.TaggedStruct("questionChoice", {
    requestId: Schema.String,
    answers: Schema.Array(AskUserQuestionAnswer),
  }),
  Schema.TaggedStruct("questionDismiss", {
    requestId: Schema.String,
  }),
  Schema.TaggedStruct("questionPark", {
    requestId: Schema.String,
  }),
  Schema.TaggedStruct("elicitAccept", {
    requestId: Schema.String,
  }),
  Schema.TaggedStruct("elicitDecline", {
    requestId: Schema.String,
  }),
  Schema.TaggedStruct("slashPick", {
    name: Schema.String,
  }),
  Schema.TaggedStruct("mentionQuery", {
    query: Schema.String,
  }),
  Schema.TaggedStruct("mentionPick", {
    path: Schema.String,
    absPath: Schema.String,
  }),
  Schema.TaggedStruct("cycleMode", {}),
  Schema.TaggedStruct("setMode", {
    modeId: Schema.String,
  }),
  Schema.TaggedStruct("setModel", {
    modelId: Schema.String,
  }),
  Schema.TaggedStruct("sendNow", {}),
  Schema.TaggedStruct("setReasoning", {
    value: Schema.String,
    modelId: Schema.optionalKey(Schema.String),
  }),
  Schema.TaggedStruct("openSettings", {}),
  Schema.TaggedStruct("openChanges", {
    turnId: Schema.optionalKey(Schema.String),
  }),
  Schema.TaggedStruct("keepAllPending", {}),
  Schema.TaggedStruct("undoAllPending", {}),
  Schema.TaggedStruct("commitAllPending", {}),
  Schema.TaggedStruct("refreshMcp", {
    name: Schema.String,
  }),
  Schema.TaggedStruct("setMcpEnabled", {
    name: Schema.String,
    enabled: Schema.Boolean,
  }),
  Schema.TaggedStruct("setMcpToolEnabled", {
    name: Schema.String,
    tool: Schema.String,
    enabled: Schema.Boolean,
  }),
  Schema.TaggedStruct("revealEditor", {
    absPath: Schema.optionalKey(Schema.String),
    startLine: Schema.optionalKey(Schema.Number),
    endLine: Schema.optionalKey(Schema.Number),
  }),
  Schema.TaggedStruct("addSelection", {}),
  Schema.TaggedStruct("openPlan", {
    markdown: Schema.optionalKey(Schema.String),
  }),
  Schema.TaggedStruct("openMcpConfig", {}),
  Schema.TaggedStruct("openSettingsJson", {}),
  Schema.TaggedStruct("trustFolder", {}),
  Schema.TaggedStruct("setSetting", {
    key: Schema.Literals([
      "cliPath",
      "nodePath",
      "includeActiveFileByDefault",
      "useCtrlEnterToSend",
      "changesPresentation",
    ]),
    value: Schema.Union([Schema.String, Schema.Boolean]),
  }),
])

export type WebviewMsg = typeof WebviewMsg.Type

export const HOST_MSG_HANDLED: Record<HostMsg["_tag"], true> = {
  ready: true,
  sessionMeta: true,
  userMessage: true,
  agentChunk: true,
  thoughtChunk: true,
  toolGroup: true,
  permissionCard: true,
  planCard: true,
  questionCard: true,
  elicitCard: true,
  availableCommands: true,
  mentionResults: true,
  composerChip: true,
  turnEnd: true,
  queued: true,
  changesSummary: true,
  editorContext: true,
  settingsState: true,
  mcpCatalog: true,
  error: true,
  clearTranscript: true,
  restoreTranscript: true,
}

export const WEBVIEW_MSG_HANDLED: Record<WebviewMsg["_tag"], true> = {
  ready: true,
  send: true,
  queue: true,
  steer: true,
  cancel: true,
  permissionChoice: true,
  permissionPark: true,
  openDiff: true,
  planVerdict: true,
  questionChoice: true,
  questionDismiss: true,
  questionPark: true,
  elicitAccept: true,
  elicitDecline: true,
  slashPick: true,
  mentionQuery: true,
  mentionPick: true,
  cycleMode: true,
  setMode: true,
  setModel: true,
  sendNow: true,
  setReasoning: true,
  openSettings: true,
  openChanges: true,
  keepAllPending: true,
  undoAllPending: true,
  commitAllPending: true,
  refreshMcp: true,
  setMcpEnabled: true,
  setMcpToolEnabled: true,
  revealEditor: true,
  addSelection: true,
  openPlan: true,
  openMcpConfig: true,
  openSettingsJson: true,
  trustFolder: true,
  setSetting: true,
}

export const HOST_MSG_TAGS = Object.keys(HOST_MSG_HANDLED) as Array<HostMsg["_tag"]>
export const WEBVIEW_MSG_TAGS = Object.keys(WEBVIEW_MSG_HANDLED) as Array<WebviewMsg["_tag"]>

export const decodeHostMsg = Schema.decodeUnknownSync(HostMsg)
export const decodeWebviewMsg = Schema.decodeUnknownSync(WebviewMsg)

File: beard/core/src/main/scala/groksbeard/core/protocol.scala
package groksbeard.core

import zio.json.*
import zio.json.ast.Json

given stringOrBooleanCodec: JsonCodec[String | Boolean] = JsonCodec(
  JsonEncoder[Json].contramap {
    case s: String  => Json.Str(s)
    case b: Boolean => Json.Bool(b)
  },
  JsonDecoder[Json].mapOrFail {
    case Json.Str(s)  => Right(s)
    case Json.Bool(b) => Right(b)
    case _            => Left("expected string or boolean")
  },
)

final case class SettingsState(
    cliPath: String,
    nodePath: String,
    includeActiveFileByDefault: Boolean,
    useCtrlEnterToSend: Boolean,
    changesPresentation: String,
) derives JsonCodec

object SettingsState:
  val defaults: SettingsState =
    SettingsState("", "", includeActiveFileByDefault = true, useCtrlEnterToSend = false, "toast")

@jsonDiscriminator("_tag")
enum HostMsg derives JsonCodec:
  @jsonHint("ready") case Ready
  @jsonHint("sessionMeta") case SessionMeta(
      sessionId: String,
      title: String,
      modeId: String,
      availableModes: List[ModeOption] = Nil,
      occupancy: Option[Occupancy] = None,
      modelId: String = "",
      availableModels: List[ModelOption] = Nil,
  )
  @jsonHint("sessionList") case SessionList(
      sessions: List[SessionRow],
      currentId: String = "",
      openPicker: Boolean = false,
  )
  @jsonHint("sessionLocked") case SessionLocked(sessionId: String, message: String)
  @jsonHint("availableCommands") case AvailableCommands(commands: List[SlashCommand])
  @jsonHint("mentionResults") case MentionResults(query: String, files: List[MentionFile])
  @jsonHint("settingsState") case Settings(
      cliPath: String,
      nodePath: String,
      includeActiveFileByDefault: Boolean,
      useCtrlEnterToSend: Boolean,
      changesPresentation: String,
  )
  @jsonHint("composerChip") case ComposerChip(
      path: String,
      absPath: String,
      source: String,
      startLine: Option[Int] = None,
      endLine: Option[Int] = None,
  )
  @jsonHint("userMessage") case UserMessage(
      turnId: String,
      text: String,
      chips: List[PromptChip] = Nil,
      steer: Boolean = false,
  )
  @jsonHint("agentChunk") case AgentChunk(turnId: String, text: String, messageId: Option[String] = None)
  @jsonHint("thoughtChunk") case ThoughtChunk(turnId: String, text: String)
  @jsonHint("toolGroup") case ToolGroup(turnId: String, tools: List[ToolRow])
  @jsonHint("permissionCard") case Permission(
      requestId: String,
      toolCallId: String,
      title: String,
      options: List[PermissionOption] = Nil,
      hasDiff: Boolean = false,
  )
  @jsonHint("planCard") case Plan(requestId: String, planMarkdown: String)
  @jsonHint("questionCard") case Question(requestId: String, questions: List[AgentQuestion])
  @jsonHint("elicitCard") case Elicit(
      requestId: String,
      serverName: String,
      mode: String,
      title: String,
      url: Option[String] = None,
  )
  @jsonHint("turnEnd") case TurnEnd(turnId: String, stopReason: String)
  @jsonHint("queued") case Queued(items: List[QueuedPrompt] = Nil)
  @jsonHint("changesSummary") case Changes(
      fileCount: Int,
      additions: Int,
      deletions: Int,
      files: List[ChangeFileView] = Nil,
  )
  @jsonHint("diffPreview") case DiffPreview(path: String, oldText: String, newText: String, wholeFile: Boolean = true)
  @jsonHint("clearDiff") case ClearDiff
  @jsonHint("error") case Error(message: String, code: Option[String] = None)
  @jsonHint("clearTranscript") case ClearTranscript
  @jsonHint("transcript") case Transcript(turns: List[TurnView] = Nil)
end HostMsg

object HostMsg:
  def settings(state: SettingsState): HostMsg =
    Settings(
      state.cliPath,
      state.nodePath,
      state.includeActiveFileByDefault,
      state.useCtrlEnterToSend,
      state.changesPresentation,
    )

  def permission(card: PermissionCard): HostMsg =
    Permission(card.requestId, card.toolCallId, card.title, card.options, card.hasDiff)

  def plan(card: PlanCard): HostMsg = Plan(card.requestId, card.planMarkdown)

  def question(card: QuestionCard): HostMsg = Question(card.requestId, card.questions)

  def elicit(card: ElicitCard): HostMsg =
    Elicit(card.requestId, card.serverName, card.mode, card.title, card.url)

  def changes(summary: ChangesSummary): HostMsg =
    Changes(summary.fileCount, summary.additions, summary.deletions, summary.files)

  def chip(p: PromptChip): HostMsg =
    ComposerChip(p.path, p.absPath, p.source, p.startLine, p.endLine)
end HostMsg

@jsonDiscriminator("_tag")
enum WebviewMsg derives JsonCodec:
  @jsonHint("ready") case Ready
  @jsonHint("send") case Send(text: String)
  @jsonHint("queue") case Queue(text: String)
  @jsonHint("cancel") case Cancel
  @jsonHint("slashPick") case SlashPick(name: String)
  @jsonHint("mentionQuery") case MentionQuery(query: String)
  @jsonHint("mentionPick") case MentionPick(path: String, absPath: String)
  @jsonHint("addSelection") case AddSelection
  @jsonHint("removeChip") case RemoveChip(
      absPath: String,
      startLine: Option[Int] = None,
      endLine: Option[Int] = None,
  )
  @jsonHint("cycleMode") case CycleMode
  @jsonHint("setMode") case SetMode(modeId: String)
  @jsonHint("setModel") case SetModel(modelId: String)
  @jsonHint("openSettings") case OpenSettings
  @jsonHint("setSetting") case SetSetting(key: String, value: String | Boolean)
  @jsonHint("permissionChoice") case PermissionChoice(requestId: String, optionId: String)
  @jsonHint("permissionPark") case PermissionPark(requestId: String)
  @jsonHint("openDiff") case OpenDiff(requestId: String)
  @jsonHint("planVerdict") case PlanVerdict(requestId: String, verdict: String)
  @jsonHint("questionChoice") case QuestionChoice(requestId: String, questionId: String, optionId: String)
  @jsonHint("questionDismiss") case QuestionDismiss(requestId: String)
  @jsonHint("elicitAccept") case ElicitAccept(requestId: String)
  @jsonHint("elicitDecline") case ElicitDecline(requestId: String)
  @jsonHint("openChanges") case OpenChanges
  @jsonHint("keepChange") case KeepChange(path: String)
  @jsonHint("undoChange") case UndoChange(path: String)
  @jsonHint("closeDiff") case CloseDiff
  @jsonHint("newSession") case NewSession
  @jsonHint("resumeSession") case ResumeSession(sessionId: String)
  @jsonHint("openSessionPicker") case OpenSessionPicker
  @jsonHint("closeSessionPicker") case CloseSessionPicker
end WebviewMsg
